lib/iolog/log_reader.go

```diff
@@ -0,0 +1,96 @@
+// Copyright 2017 VMware, Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package iolog
+
+import (
+	"encoding/binary"
+	"io"
+	"time"
+
+	log "github.com/Sirupsen/logrus"
+)
+
+// LogReader unwraps entires into LogEntry structs.
+type LogReader struct {
+	io.ReadCloser
+	prev []byte
+}
+
+// NewLogReader wraps an io.ReadCloser in a LogReader.
+func NewLogReader(r io.ReadCloser) *LogReader {
+	return &LogReader{ReadCloser: r}
+}
+
+// Read scans up to `\n` in the underlying stream, parses the header for that entry,
+// and copies the message into the supplied buffer
+func (lr *LogReader) Read(p []byte) (int, error) {
+	var (
+		err  error
+		n, w int
+	)
+	msg := lr.prev
+	if msg == nil {
+		// read a header
+		h := make([]byte, 10)
+		n = 0
+		for n < 10 {
+			w, err = lr.ReadCloser.Read(h[n:])
+			n += w
+			if err != nil {
+				break
```

Comment:
looks like you can return from here, no need to break from the loop, check one more time and then return